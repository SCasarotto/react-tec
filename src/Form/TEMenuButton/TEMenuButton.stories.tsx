import React from 'react'
import { Story, Meta } from '@storybook/react'

import { TEMenuButton, TEMenuButtonProps } from './TEMenuButton'
import { TEButton } from '../TEButton'

export default {
	title: 'Form/TEMenuButton',
	component: TEMenuButton,
	argTypes: {},
} as Meta

const Template: Story<TEMenuButtonProps> = (args) => <TEMenuButton {...args} />

export const Default = Template.bind({})
Default.args = {
	title: 'Click Me',
	children: ({ hideMenu }) => {
		return (
			<div>
				<TEButton
					onClick={() => {
						alert('Button 1 Clicked')
						hideMenu()
					}}
				>
					Button 1
				</TEButton>
				<TEButton
					onClick={() => {
						alert('Button 2 Clicked')
						hideMenu()
					}}
				>
					Button 2
				</TEButton>
				<TEButton
					onClick={() => {
						alert('Button 3 Clicked')
						hideMenu()
					}}
				>
					Button 3
				</TEButton>
			</div>
		)
	},
}
