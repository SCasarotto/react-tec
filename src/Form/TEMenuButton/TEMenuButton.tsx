import React, { useState, useRef, useCallback, useEffect, ReactNode } from 'react'
import { Reference, Popper, PopperProps, PopperChildrenProps } from 'react-popper'

import { TEButton } from './../TEButton'

export interface TEMenuButtonChildrenProps extends PopperChildrenProps {
	hideMenu(): void
}
//TODO: Improve Types
export interface TEMenuButtonProps extends Omit<PopperProps<any>, 'children'> {
	className?: string
	title: ReactNode
	children(props: TEMenuButtonChildrenProps): ReactNode
}
/**
 *
 * TEMenuButton is a simple dropdown component. This component relies primarily on [React Popper](https://github.com/popperjs/react-popper) Component.
 *
 * All the props and types are the same besides having added `hideMenu` function to the data that is returned to the children.
 *
 * Please note that in order for this component to function, the app must be wrapped by `TEAppWrapper` because it contains the `Manager` wrapping component required by Popper.
 *
 */
export const TEMenuButton: React.FC<TEMenuButtonProps> = (props) => {
	const { placement, className = '', title, children, ...rest } = props
	const [menuVisible, setMenuVisible] = useState(false)
	const containerRef = useRef(null)
	const hideMenu = useCallback(() => setMenuVisible(false), [])
	const handleOutsideClick = useCallback((e) => {
		// @ts-ignore: Object is possibly 'null'.
		if (containerRef && containerRef.current && containerRef.current.contains(e.target)) {
			return
		}
		setMenuVisible(false)
		return
	}, [])
	useEffect(() => {
		if ('ontouchend' in window) {
			document.addEventListener('touchend', handleOutsideClick)
		} else {
			document.addEventListener('click', handleOutsideClick)
		}
		return () => {
			document.removeEventListener('touchend', handleOutsideClick)
			document.removeEventListener('click', handleOutsideClick)
		}
	}, [handleOutsideClick])

	return (
		<div className={`TEMenuButton ${className}`} ref={containerRef}>
			<Reference>
				{({ ref }) => (
					<TEButton
						className='TEMenuButtonButton'
						ref={ref}
						onClick={() => setMenuVisible(!menuVisible)}
					>
						{title}
					</TEButton>
				)}
			</Reference>
			<Popper placement={placement} {...rest}>
				{(popperData) =>
					menuVisible ? (
						<div
							className='TEMenuButtonMenu'
							ref={popperData.ref}
							style={popperData.style}
							data-placement={popperData.placement}
						>
							{children({ ...popperData, hideMenu })}
						</div>
					) : null
				}
			</Popper>
		</div>
	)
}
