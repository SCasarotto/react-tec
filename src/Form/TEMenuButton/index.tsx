import React from 'react'

import { Button } from './styledComponents'

export interface TEMenuButtonProps
	extends React.DetailedHTMLProps<
		React.ButtonHTMLAttributes<HTMLButtonElement>,
		HTMLButtonElement
	> {
	//TODO: Fix this. styled-components is weird with refs
	ref?: any
}
export const TEMenuButton: React.FC<TEMenuButtonProps> = (props) => {
	const { className = '', type = 'button', ...rest } = props
	return <Button className={`TEButton ${className}`} type={type} {...rest} />
}
