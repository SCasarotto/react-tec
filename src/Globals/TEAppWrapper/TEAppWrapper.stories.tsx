import React from 'react'
import { Story, Meta } from '@storybook/react'

import { TEAppWrapper, TEAppWrapperProps } from './TEAppWrapper'

export default {
	title: 'Globals/TEAppWrapper',
	component: TEAppWrapper,
	argTypes: {},
} as Meta

const Template: Story<TEAppWrapperProps> = (args) => <TEAppWrapper {...args} />

export const Default = Template.bind({})
Default.args = {}
