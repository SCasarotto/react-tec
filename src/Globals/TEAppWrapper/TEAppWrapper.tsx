import React from 'react'
import { BrowserRouter, useLocation } from 'react-router-dom'
import { Manager } from 'react-popper'

import { useScrollTop } from '../../Utility/useScrollTop'
import { TEThemeProvider, TEThemeProviderProps } from '../TEThemeContext'
import { TEPopupProvider } from '../TEPopupContext'

export interface TEAppWrapperProps extends TEThemeProviderProps {}
const TEAppWrapperContent: React.FC<TEAppWrapperProps> = (props) => {
	const { theme, globalStyles, children } = props
	const location = useLocation()
	useScrollTop(location)

	return (
		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
			<TEPopupProvider>
				<Manager>{children}</Manager>
			</TEPopupProvider>
		</TEThemeProvider>
	)
}

/**
 *
 * TEAppWrapper is meant to be the single top level wrapping element you need. Having said that, it does contain a number of things. `TEAppWrapper` contain the following:
 *
 * ```
 * <BrowserRouter>
 * 	<TEScrollToTop>
 * 		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
 *			<TEPopupProvider>{children}</TEPopupProvider>
 *		</TEThemeProvider>
 * 	</TEScrollToTop>
 * </BrowserRouter>
 * ```
 *
 * Here `BrowserRouter` comes from [react-router-dom](https://reacttraining.com/react-router/web/api/BrowserRouter). For all other components, please view thier given documentation page.
 *
 * As for props, the only props `TEAppWrapper` accepts are those that `TEThemeProvider` accepts.
 *
 * `TEAppWrapper` is implemented like the following in this project:
 *
 * ```
 * const Main = () => (
 * 	<TEAppWrapper globalStyles={globalStyles}>
 *		<App />
 * 	</TEAppWrapper>
 * )
 * ReactDOM.render(<Main />, document.getElementById('root'))
 * ```
 *
 */
export const TEAppWrapper: React.FC<TEAppWrapperProps> = (props) => {
	return (
		<BrowserRouter>
			<TEAppWrapperContent {...props} />
		</BrowserRouter>
	)
}
