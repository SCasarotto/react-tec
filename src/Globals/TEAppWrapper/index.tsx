import React from 'react'
import {
	BrowserRouter,
	RouteComponentProps,
	useLocation,
} from 'react-router-dom'

import { useScrollTop } from '../../Utility/useScrollTop'
import { TEThemeProvider, TEThemeProviderProps } from '../TEThemeContext'
import { TEPopupProvider, TEPopupProviderProps } from '../TEPopupContext'
interface TEAppWrapperProps
	extends TEThemeProviderProps,
		TEPopupProviderProps,
		RouteComponentProps {}

const TEAppWrapperContent: React.FC<TEAppWrapperProps> = (props) => {
	const { theme, globalStyles, children } = props
	const location = useLocation()
	useScrollTop(location)

	return (
		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
			<TEPopupProvider>{children}</TEPopupProvider>
		</TEThemeProvider>
	)
}

export const TEAppWrapper: React.FC<TEAppWrapperProps> = (props) => {
	return (
		<BrowserRouter>
			<TEAppWrapperContent {...props} />
		</BrowserRouter>
	)
}
