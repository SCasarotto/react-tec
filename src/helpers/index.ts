export * from './colorHelpers'
export * from './manipulateRowData'
