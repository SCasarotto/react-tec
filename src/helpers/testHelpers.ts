export const mockOnScroll = () => {
	const { scrollTo } = window
	beforeAll(() => {
		// Replace with custom value
		window.scrollTo = jest.fn()
	})
	afterAll(() => {
		// Restore original
		window.scrollTo = scrollTo
	})
}
