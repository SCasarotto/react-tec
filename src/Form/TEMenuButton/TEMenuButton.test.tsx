import React from 'react'
import { render } from '@testing-library/react'
import { TEMenuButton } from './TEMenuButton'
import { TEAppWrapper } from './../../Globals/TEAppWrapper'
import { mockOnScroll } from '../../helpers/testHelpers'

describe('TEMenuButton', () => {
	mockOnScroll()
	it('matches snapshot', () => {
		const { asFragment } = render(
			<TEAppWrapper>
				<TEMenuButton title='My Menu Button'>
					{({ hideMenu }) => (
						<div>
							<button
								onClick={() => {
									alert('Press!1')
									hideMenu()
								}}
							>
								My Button1
							</button>
							<button
								onClick={() => {
									alert('Press!2')
									hideMenu()
								}}
							>
								My Button2
							</button>
							<button
								onClick={() => {
									alert('Press!3')
									hideMenu()
								}}
							>
								My Button3
							</button>
						</div>
					)}
				</TEMenuButton>
			</TEAppWrapper>,
		)
		expect(asFragment()).toMatchSnapshot()
	})
})
