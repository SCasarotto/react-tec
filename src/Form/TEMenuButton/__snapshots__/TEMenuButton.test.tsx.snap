// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEMenuButton matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TEMenuButton "
  >
    <button
      class="sc-bdnylx iREeCR TEButton TEMenuButtonButton"
      type="button"
    >
      My Menu Button
    </button>
  </div>
  <div
    class="sc-gtssRu kELAqg TEPopup sc-hKFyIo eFASRH TEAlert "
  >
    <div
      class="sc-dlnjPT eygYex TEPopupContent"
    >
      <button
        class="sc-bdnylx iREeCR TEButton sc-gKAblj eCNrJN TEAlertButton"
        type="button"
      >
        Okay
      </button>
    </div>
  </div>
  <div
    class="sc-gtssRu kELAqg TEPopup sc-iCoHVE bPtAHf TEConfirm "
  >
    <div
      class="sc-dlnjPT eygYex TEPopupContent"
    >
      <div
        class="sc-jrsJCI eaZarr TEConfirmButtonoContainer"
      >
        <button
          class="sc-bdnylx iREeCR TEButton sc-kEqYlL kGJRdG TEConfirmButton TEConfirmLeftButton"
          type="button"
        >
          Cancel
        </button>
        <button
          class="sc-bdnylx iREeCR TEButton sc-iqAbSa fiTZwB TEConfirmButton TEConfirmRightButton"
          type="button"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-gtssRu kELAqg TEPopup sc-ksluoS conrax TENetworkActivity "
  >
    <div
      class="sc-dlnjPT eygYex TEPopupContent"
    >
      <div
        class="sc-crzoUp cspjma TESpinner sc-hBMVcZ dhWjui TENetworkActivitySpinner"
      >
        <div
          class="sc-dIsAE ibfIYE TESpinnerContainer"
        >
          <div
            class="sc-bqGHjH jxGHGS TESpinnerRing TESpinnerRing1"
          />
          <div
            class="sc-bqGHjH jhQThV TESpinnerRing TESpinnerRing2"
          />
          <div
            class="sc-bqGHjH gnowos TESpinnerRing TESpinnerRing3"
          />
          <div
            class="sc-bqGHjH fJuDOy TESpinnerRing TESpinnerRing4"
          />
          <div
            class="sc-bqGHjH kESRgu TESpinnerRing TESpinnerRing5"
          />
          <div
            class="sc-bqGHjH bZa-DvO TESpinnerRing TESpinnerRing6"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
